flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RecoveredInputChannelTest.java

```diff
@@ -220,24 +223,24 @@ private void testConcurrentReadStateAndProcessAndRelease(boolean isRemote) throw
 		}
 	}
 
-	private Callable<Void> readRecoveredStateTask(RecoveredInputChannel inputChannel, ChannelStateReader reader) {
+	private Callable<Void> readRecoveredStateTask(RecoveredInputChannel inputChannel, ChannelStateReader reader, boolean verifyRelease) {
 		return () -> {
 			try {
 				inputChannel.readRecoveredState(reader);
-			} catch (CancelTaskException ex) {
-				// within expectation
+			} catch (Throwable t) {
+				assertTrue("The expected exception should only happen in the case of released channel.", verifyRelease && inputChannel.isReleased());
```

Comment:
Just to clarify: do we expect an exception to be thrown in this case or is it a possible (undeterministic) outcome?
If we expect it, then the test needs to be modified.
If it’s undeterministic, is there a way to make it deterministic?

Reply:
The previous expected `CancelTaskException` happens inside `BufferManager#requestBufferBlocking` while reading recovered state by IO thread. In strict way, I think it also makes sense to limit the exception here as before.

Regarding the concern of deterministic outcome, actually the determined outcome of this test are no buffer leaks and deadlock issues finally, but not for the determined thrown exception case. 

If we want to strictly control the executions among IO thread and canceler thread, that means before IO thread requesting the floating buffer, then the canceler thread already released the channel to give a determined thrown exception. But it only covers the partial scenarios of the concurrent execution. In practice, these two threads can execute concurrently in any order, e.g. after IO thread already requested the floating buffer, the canceler thread works. So we did not make the precise control of them, just make them execute in random way which is more nearly in real scenario and can cover all the possible cases. 

Actually we had many similar cases in `RemoteInputChannelTest` to detect the potential deadlock issues among three threads.

Reply:
I'm still concerned that the test is not as specific as it could be.
Could we assume that an exception needs to be thrown at either place? (there is a second try-catch block) Or could it happen that no exception is thrown at all?

If this is non-deterministic, would it make sense to repeatedly execute the test? The test is rather fast on my machine.

Reply:
I'm still not convinced about this test. It's reimplementing production code in many aspects. For example exception handling (this exception ignoring during cancelation/closing) and threading model from `Task`/`StreamTask`, so at any time we modify those, this test has a chance to fail. 

Also I'm pretty sure there are `ITCase`s testing for that and the logic there is simpler and more stable manner.

Reply:
> Could we assume that an exception needs to be thrown at either place? (there is a second try-catch block) Or could > it happen that no exception is thrown at all?

Yeah, I understand your concern. Actually this test covers the following scenarios:
1. First: IO thread inserts the buffer into channel queue, Second: canceler thread releases the channel to recycle the queued buffer with exception. Without exception in this case.

2. First: Canceler thread releases the channel, Second: IO thread tries to insert the buffer into channel queue. But it would check the released state and then recycle the filled buffer by IO thread. Without exception in this case.

3. First: IO thread consumes all the available buffers while reading states and blocks on waiting for buffers. Second: canceler thread releases the channel and awake the IO thread. Then IO thread will check the isReleased state and then throw `CancelTaskException`.

4. First: Canceler thread releases the channel, Second: IO thread requests buffer but does not get available buffer because of released channel, then it will check the `isReleased` state and throw `CancelTaskException`. The difference with above is that the IO thread is not awaked by canceler thread.

Maybe there are also some other cases besides above. So if we want to get determined results, we need to design different separate tests for fine-grained scenarios, and I am not quite sure whether it is easy to control the logics among different processes. And it might be more fragile future if we change the related processes because we touch the fine-grained logics.

From another aspect, the overall scenario is that while IO thread recovering state, the canceler thread is releasing the channel. As long as we construct this scenario, we do not need to touch the fine-grained logics inside every process, somehow like the ITCase as @pnowojski said above. And finally we only need to guarantee there are no deadlock and buffer leaks in the result and ignores the intermediate results.  

> If this is non-deterministic, would it make sense to repeatedly execute the test? The test is rather fast on my machine.

I think it is hard to give a proper loop number to guarantee which action will be executed among these threads, then it might cause more fragile cases. Actually somehow we rely on this non-determined environment to try to cover all the possible cases, as mentioned in above four cases.

Reply:
Okay thank you for the detailed explanation. I know understood the challenges. I'd also second @pnowojski that the tests are probably too fine-grain.

Reply:
> I'm still not convinced about this test. It's reimplementing production code in many aspects. For example exception handling (this exception ignoring during cancelation/closing) and threading model from Task/StreamTask, so at any time we modify those, this test has a chance to fail.

I guess the reimplementing production code refers to `processRecoveredBufferTask` which will consumed the recovered buffers and verify the data. This test actually only focuses on `RecoveredInputChannel#readRecoveredState` method, so as long as the related interaction processes is not changed, this test is still valid, no matter with how the upper component handles the `CancelTaskException` (ignore or throw it) .

In general the unit tests should avoid touching more components to bring potential fragile if some changes future. From this aspect, maybe the ITCase is more suitable for this scenario as you said. But unit tests still have somehow meanings to work on the respective methods directly, then it might has more chances to cover all the potential cases.

Reply:
> I'm still not convinced about this test. It's reimplementing production code in many aspects. For example exception handling (this exception ignoring during cancelation/closing) and threading model from Task/StreamTask, so at any time we modify those, this test has a chance to fail.

I guess the reimplementing production code refers to processRecoveredBufferTask which will consumed the recovered buffers and verify the data. This test actually only focuses on RecoveredInputChannel#readRecoveredState method, so as long as the related interaction processes is not changed, this test is still valid, no matter with how the upper component handles the CancelTaskException (ignore or throw it) .

In general the unit tests should avoid touching more components to bring potential fragile if some changes future. From this aspect, maybe the ITCase is more suitable for this scenario as you said. But unit tests still have somehow meanings to work on the respective methods directly, then it might has more chances to cover all the potential cases.